File: backend/src/main/java/es/udc/paproject/backend/model/services/AnimalService.java
package es.udc.paproject.backend.model.services;

import es.udc.paproject.backend.model.entities.Animal;
import es.udc.paproject.backend.model.exceptions.InstanceNotFoundException;
import es.udc.paproject.backend.model.exceptions.PermissionException;
import es.udc.paproject.backend.rest.dtos.FoodPurchaseDTOs.ConsumptionChartDto;
import org.hibernate.tuple.AbstractNonIdentifierAttribute;

import java.util.List;

public interface AnimalService {

    void createAnimal(Animal animal);

    void updateAnimal(Long id, Animal animal) throws InstanceNotFoundException;

    void setAnimalDead(Long userId, Long id) throws InstanceNotFoundException, PermissionException;

    Animal getAnimalById(Long userId, Long id) throws InstanceNotFoundException;

    Block<Animal> getAllAnimals(Long userId, String name, Long identificationNumber, String startDate, String endDate,
        Boolean isMale, int page, int size) throws InstanceNotFoundException;

    List<Animal> getAnimalsWithLabel(Long userId, boolean all, boolean onlyFemale) throws InstanceNotFoundException;

    List<ConsumptionChartDto> getAnimalFoodConsumptionChart(Long userId, Long animalId) throws InstanceNotFoundException;

}

File: frontend/src/backend/animalService.js
import {config, appFetch} from "./appFetch";

export const createAnimal = (animal, onSuccess, onErrors) => {
    appFetch('/animal/registerAnimal', config('POST', animal),
        onSuccess, onErrors);
}

export const updateAnimal = (animalId, animal, onSucces, onErrors) => {
    appFetch(`/animal/${animalId}`, config('PUT', animal),
        onSucces, onErrors);
}

export const getAnimalById = (animalId, onSuccess, onErrors) => {
    appFetch(`/animal/${animalId}`, config('GET'),
        onSuccess, onErrors);
}

export const deleteAnimal = (animalId, onSuccess, onErrors) => {
    appFetch(`/animal/${animalId}/isDead`, config('DELETE'),
        onSuccess, onErrors);
}

export const getAllAnimals = ({page, name, identificationNumber, startDate, endDate, isMale, dead}, onSuccess, onErrors) => {
    let path = `/animal/allAnimals?page=${page}&size=5`;

    path += name?.length > 0 ? `&name=${encodeURIComponent(name)}` : "";
    path += identificationNumber ? `&identificationNumber=${encodeURIComponent(identificationNumber)}` : "";
    path += startDate ? `&startDate=${encodeURIComponent(startDate.toISOString().substring(0,10))}` : "";
    path += endDate ? `&endDate=${encodeURIComponent(endDate.toISOString().substring(0,10))}` : "";
    path += isMale ? `&isMale=${encodeURIComponent(isMale)}` : "";
    path += dead ? `&dead=${encodeURIComponent(dead)}` : "";

    appFetch(path, config('GET'), onSuccess, onErrors);
}

export const getAllAnimalsWithLabel = (all, onlyFemale, onSuccess, onErrors) => {
    if (all && onlyFemale) {
        appFetch('/animal/animalsWithLabel?all=true&onlyFemale=true', config('GET'), onSuccess, onErrors);
    } else if (all) {
        appFetch('/animal/animalsWithLabel?all=true', config('GET'), onSuccess, onErrors);
    } else if (onlyFemale) {
        appFetch('/animal/animalsWithLabel?onlyFemale=true', config('GET'), onSuccess, onErrors);
    } else {
        appFetch('/animal/animalsWithLabel', config('GET'), onSuccess, onErrors);
    }
}

export const getAnimalFoodConsumptionChartData = (animalId, onSuccess, onErrors) => {
    appFetch(`/animal/animalFoodConsumptionChart/${animalId}`, config('GET'), onSuccess, onErrors);
}
